File: Challenge-2023-01/22-palindrome-partitioning.cpp
#include "../header.h"

class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> res;
        vector<string> current;
        helper(s, res, current, 0);
        return res;
    }

    bool isPalindrome(const string& s, int l, int r) {
        while (l < r) {
            if (s[l] != s[r]) {
                return false;
            }
            ++l;
            --r;
        }
        return true;
    }

    void helper(const string& s, vector<vector<string>>& res, vector<string>& current, int index) {
        if (index == s.size()) {
            res.push_back(current);
        } else {
            for (int i = index; i < s.size(); ++i) {
                if (isPalindrome(s, index, i)) {
                    current.push_back(s.substr(index, i - index + 1));
                    helper(s, res, current, i + 1);
                    current.pop_back();
                }
            }
        }
    }
};

File: Challenge-2024-05/22-palindrome-partitioning.py
from typing import List


class Solution:
    def partition(self, s: str) -> List[List[str]]:
        def isPalindrome(start: int, end: int) -> bool:
            while start < end:
                if s[start] == s[end]:
                    start += 1
                    end -= 1
                else:
                    return False
            return True

        def helper(res: List[List[str]], current: List[str], index: int) -> None:
            if index == len(s):
                tmp = []
                for x in current:
                    tmp.append(x)
                res.append(tmp)
                return
            else:
                for i in range(index, len(s)):
                    if isPalindrome(index, i):
                        current.append(s[index:i+1])
                        helper(res, current, i+1)
                        current.pop()
        res, current = [], []
        helper(res, current, 0)
        return res
    
if __name__ == '__main__':
    s = 'aab'
    res = Solution().partition(s)
    print(res)